Support the Calendar API

buckett opened the pull request:
This adds support for calendar events. It's not quite finished so don't feel obliged to review, but I thought it was worth putting it in a PR incase you had any feedback or if anyone wanted to build on it before I do.

ToeBee commented:
Thanks for the PR. Let's see what Jenkins thinks of it before I dive in too deep :)

Jenkins: ok to test.

ToeBee commented:
One of your new tests seems to be failing. Here is the output from Jenkins:

```
Running edu.ksu.canvas.tests.calendar.CalendarReaderUTest
INFO  [edu.ksu.canvas.net.FakeRestClient            ] adding url to response map: domain/api/v1/calendar_events
DEBUG [edu.ksu.canvas.impl.CalendarEventImpl        ] List calendar events for current user
DEBUG [edu.ksu.canvas.util.CanvasURLBuilder         ] Built Canvas url - domain/api/v1/calendar_events
DEBUG [edu.ksu.canvas.impl.RestCanvasMessenger      ] Sending GET request to: domain/api/v1/calendar_events
DEBUG [edu.ksu.canvas.impl.RestCanvasMessenger      ] Sending GET request to: domain/api/v1/calendar_events
DEBUG [edu.ksu.canvas.net.FakeRestClient            ] Sending fake GET to domain/api/v1/calendar_events
INFO  [edu.ksu.canvas.net.FakeRestClient            ] adding url to response map: domain/api/v1/calendar_events/95
INFO  [edu.ksu.canvas.impl.CalendarEventImpl        ] Getting calendar event: 95
DEBUG [edu.ksu.canvas.util.CanvasURLBuilder         ] Built Canvas url - domain/api/v1/calendar_events/95
DEBUG [edu.ksu.canvas.impl.RestCanvasMessenger      ] Sending GET request to: domain/api/v1/calendar_events/95
DEBUG [edu.ksu.canvas.net.FakeRestClient            ] Sending fake GET to domain/api/v1/calendar_events/95
INFO  [edu.ksu.canvas.net.FakeRestClient            ] adding url to response map: domain/api/v1/users/73/calendar_events
INFO  [edu.ksu.canvas.impl.CalendarEventImpl        ] List calendar events for 73
DEBUG [edu.ksu.canvas.util.CanvasURLBuilder         ] Built Canvas url - domain/api/v1/users/73/calendar_events
DEBUG [edu.ksu.canvas.impl.RestCanvasMessenger      ] Sending GET request to: domain/api/v1/users/73/calendar_events
DEBUG [edu.ksu.canvas.impl.RestCanvasMessenger      ] Sending GET request to: domain/api/v1/users/73/calendar_events
DEBUG [edu.ksu.canvas.net.FakeRestClient            ] Sending fake GET to domain/api/v1/users/73/calendar_events
INFO  [edu.ksu.canvas.impl.CalendarEventImpl        ] Getting calendar event: 1
DEBUG [edu.ksu.canvas.util.CanvasURLBuilder         ] Built Canvas url - domain/api/v1/calendar_events/1
DEBUG [edu.ksu.canvas.impl.RestCanvasMessenger      ] Sending GET request to: domain/api/v1/calendar_events/1
DEBUG [edu.ksu.canvas.net.FakeRestClient            ] Sending fake GET to domain/api/v1/calendar_events/1
Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.028 sec <<< FAILURE! - in edu.ksu.canvas.tests.calendar.CalendarReaderUTest
testGetCalendarEventMissing(edu.ksu.canvas.tests.calendar.CalendarReaderUTest)  Time elapsed: 0.008 sec  <<< FAILURE!
java.lang.AssertionError: Expected exception: java.io.IOException
```

buckett commented:
I've cleaned up the commits, but I can't manage to reproduce the build failure. It always builds/test fine for me on this branch.

buckett commented:
The test failure was only happening on Linux, I've managed to reproduce it and fix it.

ToeBee commented:
Some of the new Time classes were giving me pause. But I guess everything that is still using `java.util.Date` will still use the existing type adapter and the things you added that uses `Instant` will use the new adapter so that should be safe. I guess we should have gone with more `java.time` stuff originally since this library is specifically for Java 8+.

ToeBee commented:
Ok, a couple of questions/comments:
* Should the calendar event ID being passed in to the methods like `deleteCalendarEvent` be a String or an Integer? In the CalendarEvent object the ID is an Integer. Strings have to be used for IDs that can be referred to by SIS ID (so courses, users, a few others) but here it seems like they could be integers.
* If you are requiring an action options class to be passed in, you could put the required object ID in the options class and pass it in as a required constructor parameter. For example how it happens in [GetSubAccountsOptions](https://github.com/kstateome/canvas-api/blob/master/src/main/java/edu/ksu/canvas/requestOptions/GetSubAccountsOptions.java). So for example  
`deleteCalendarEvent("123", new DeleteCalendarEventOptions())`  
would turn into  
`deleteCalendarEvent(new DeleteCalendarEventOptions("123"))`  
* Since the `context_code` is required to create an event, we should probably throw an exception (IllegalArgumentException?) if it isn't present so that you don't get a weird 404 error (which I agree seems like a bug in Canvas). The combination of required information is odd. You can create an event with just context_code and title or just context_code and a start date. But with *just* a context code it throws a 400 error. 
* `allDay` should probably not default to `true` in the CalendarEvent object since nothing else has default values.

ToeBee commented:
I don't think the `CreateCalendarEventOptions` class is ever used? The creation method takes in a `CalendarEvent` object and serializes it via the `toPostMap()` method. Which... I'm guessing doesn't correctly handle the duplication and child event fields correctly. And that child event stuff actually looks kind of problematic with the way things are currently set up. So I'm not going to ask that you get that working in this PR.  But maybe add something in comments saying that those features are not complete yet.

As for the failing build: It looks like Jenkins is trying to merge this branch with master before building it and while the merge works, I made some changes in some base classes that break compilation. It was to correctly support the "serialize nulls" option when using the form POST method of talking to Canvas as opposed to the JSON POST. You can try merging master into your branch or I can probably fix it pretty easily after this PR gets merged.

buckett commented:
I've removed the unused files (options ones).
I've also added some integration tests to support this.
Duplicate events is working (I think).
Events for specific sections are working (I think).
I've merged in master to there shouldn't be any conflicts.

ToeBee commented on README.md:
```diff
@@ -85,5 +85,24 @@ If this code is run on a course with 325 users it would print the following:
+
+To run the integration tests use:
+
+    mvn -Pfailsafe test
```
The integration tests didn't run until I changed this to `verify` which I think is the default goal for failsafe.
```suggestion
    mvn -Pfailsafe verify
```

ToeBee commented:
Awesome, you added some integration tests! I did run into a problem when running them though. 

The integration test is supposed to create a daily recurring event but it ends up actually being a weekly event. The problem is that the string `DAILY` is being sent over the wire instead of `daily` and Canvas silently ignores non matching values and changes it to weekly instead.

I see you are using the `@SerializedName` annotation on the enums in the CalendarEvent object to try and set this value. However this annotation is only used by Gson and CalendarEventImpl is using the form POST method to serialize the values over the wire, not a JSON POST. It looks like the way you implemented the child events requires using the form POST so we can't just switch to JSON. I was looking at how to get child events to work with JSON and it looked like a pretty big problem so I think your approach with using the form POST is good, we just need to figure out how to get it to serialize out the correct values for the enums.

The easy fix would be to make the enums implement `toString()` which could be as simple as:
```
@Override
public String toString() { return name().toLowerCase(); }
```
The other option would be to make the `BaseCanvasModel.toPostMap()` method respect the `@SerializedName` annotation. This one would actually let other enums that use the toString approach to be converted to use the annotation which might be nice.

I have to go catch a bike ride right now but I will work on this some more tomorrow. I might go ahead and merge this PR and then fix the enum problem one way or the other.
